swupd/delta.go

```diff
@@ -0,0 +1,186 @@
+package swupd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/pkg/errors"
+)
+
+// CreateDeltas creates all delta files between the previous and current
+// version of the supplied manifest, returning a list of files that
+// delta creation failed for.
+func CreateDeltas(manifest, statedir string, from, to uint32) ([]*File, error) {
+	var c config
+
+	c, err := getConfig(statedir)
+	if err != nil {
+		return nil, err
+	}
+
+	var oldManifest *Manifest
+	var newManifest *Manifest
+
+	if oldManifest, err = ParseManifestFile(filepath.Join(c.outputDir, fmt.Sprintf("%d", from), manifest)); err != nil {
+		return nil, err
+	}
+	if newManifest, err = ParseManifestFile(filepath.Join(c.outputDir, fmt.Sprintf("%d", to), manifest)); err != nil {
+		return nil, err
+	}
+
+	// Must be sorted before passing into linkPeersAndChange()
+	oldManifest.sortFilesName()
+	newManifest.sortFilesName()
+
+	_, _, _ = newManifest.linkPeersAndChange(oldManifest, from)
+
+	deltas, err := consolidateDeltaFiles(newManifest, from, c)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to create deltas list %s", manifest)
+	}
+
+	const maxRoutines = 3
+	var failed []*File
+	var deltaQueue = make(chan *File)
+	var failedChan = make(chan *File)
+	var wg sync.WaitGroup
+	wg.Add(maxRoutines)
+
+	for i := 0; i < maxRoutines; i++ {
+		go func(statedir string) {
```

Comment:
Without having tried it, I suspect you can remove statedir from here and line 62, Then line 57 could pick it up from line 16. There are lots of good reasons for passing in things explicitly, in particular if you think they are going to change during the lifetime of the goroutine, but I don't think this is one of them.